js/curation-tool.js

```diff
@@ -413,10 +464,308 @@ var vm = new Vue({
+            // a given phylogeny.
+            // Wrapper for does_specifier_match_node() on every node label in
+            // a given phylogeny.
+
+            let local_vm = this;
+
+            return local_vm.get_node_labels_in_phylogeny(phylogeny).filter(function(node_label) {
+                return local_vm.does_specifier_match_node(specifier, phylogeny, node_label);
+            });
+        },
+
+        does_specifier_match_node: function(specifier, phylogeny, node_label) {
+            // Tests whether a specifier matches a node in a phylogeny.
+
+            // Does the specifier have any taxonomic units? If not, we can't
+            // match anything!
+            if(!specifier.hasOwnProperty('referencesTaxonomicUnits'))
+                return false;
+
+            // Find all the taxonomic units associated with the specifier and
+            // with the node.
+            let specifier_tunits = specifier.referencesTaxonomicUnits;
+            let node_tunits = this.get_tunits_for_node_label_in_phylogeny(phylogeny, node_label);
+
+            // Attempt pairwise matches between taxonomic units in the specifier
+            // and associated with the node.
+            for(let tunit1 of specifier_tunits) {
+                for(let tunit2 of node_tunits) {
+                    if(
+                        do_tunits_match_by_binomial_name(tunit1, tunit2) ||
+                        do_tunits_match_by_external_references(tunit1, tunit2) ||
+                        do_tunits_match_by_specimen_identifier(tunit1, tunit2)
+                    ) return true;
+                }
+            }
+
+            return false;
         }
     }
 });
 
+// The following functions test whether a pair of taxonomic units match
+// given certain properties. Eventually, we will encapsulate taxonomic units
+// into their own Javascript class; once we do, these functions will become
+// methods in that class.
+
+/**
+ * do_tunits_match_by_external_references(tunit1, tunit2)
+ *
+ * Test whether two Tunits have matching external references.
+ */
+function do_tunits_match_by_external_references(tunit1, tunit2) {
+    if(tunit1 === undefined || tunit2 === undefined) return false;
+    if(tunit1.hasOwnProperty('externalReferences') && tunit2.hasOwnProperty('externalReferences')) {
+        // Each external reference is a URL as a string. We will lowercase it,
+        // but do no other transformation.
+        for(let extref1 of tunit1.externalReferences) {
+            for(let extref2 of tunit2.externalReferences) {
+                if(extref1.trim() != '' && extref1.toLowerCase().trim() === extref2.toLowerCase().trim())
+                    return true;
+            }
+        }
+    }
+
+    return false;
+}
+
+/**
+ * do_scnames_match_by_binomial_name(scname1, scname2)
+ *
+ * Test whether two scientific names match on the basis of their binomial names.
+ */
+function do_scnames_match_by_binomial_name(scname1, scname2) {
```

Comment:
Functions that return a boolean generally won't stand on their own when called. Hence, giving them names that constitute a full sentence make not only for awkward reading when they're invoked (`if (do_sth_match_by_sth())`) but also can be easily misread as an action (`if (do_something)`).

Maybe the better model here is having a TU class and a ScientificName class, and then give them methods such as `tu.matches_TU()` and `tu.matches_scientificName`.

BTW camelCase is generally far more common in JS than snake_case. I'd suggest you use the style that's prevalent in the respective language. I.e., don't make JS look like Python.